patches/api/0416-Add-whitelist-events.patch

```diff
@@ -0,0 +1,124 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SageSphinx63920 <sage@sagesphinx63920.dev>
+Date: Sun, 14 May 2023 12:56:15 +0200
+Subject: [PATCH] Add whitelist events
+
+
+diff --git a/src/main/java/io/papermc/paper/event/player/PlayerWhitelistAddEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerWhitelistAddEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..7d788bdd52d63f96c56f0a13cddd6101d29e8dc7
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/player/PlayerWhitelistAddEvent.java
+@@ -0,0 +1,53 @@
++package io.papermc.paper.event.player;
++
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.UUID;
++
++/**
++ * Called when a player is being added to the whitelist
++ */
++public class PlayerWhitelistAddEvent extends Event implements Cancellable {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++    private boolean cancel = false;
++    @NotNull private final UUID who;
++
++    public PlayerWhitelistAddEvent(@NotNull UUID who) {
++        this.who = who;
++    }
++
++    /**
++     * Gets the UUID of the player being added to the whitelist
++     *
++     * @return the UUID of the player being added to the whitelist
++     */
++    @NotNull
++    public UUID getUUID() {
```

Comment:
The rest of the whitelist API, e.g. `Server#getWhitelistedPlayers` uses `OfflinePlayer`s. Those should probably also be used here?

Reply:
The method i used for triggering the event only return a gameprofile and i was unsure if i could use Bukkit.getPlayer. I would let the plugins handle how they use this uuid because creating an offline player would do extra server load

Reply:
Getting an offline player via the `CraftServer#getOfflinePlayer(GameProfile)` method won't cause any unnecessary server load since it doesn't do any lookups, you're likely confusing it with the String method that does.

Reply:
So i should use this `if (!new io.papermc.paper.event.player.PlayerWhitelistRemoveEvent(org.bukkit.Bukkit.getOfflinePlayer(profile.getId())).callEvent()) return;`?